[13.0] stock_storage_type & stock_storage_type_putaway_abc: handle max_height

sebalix opened the pull request:
**1.  Module `stock_storage_type`:**

New `max_height` field on the the location storage type to indicate the max height supported by a location among all its storage type.

We want to propose the first location corresponding to the height of a package by losing a minimum of space, so we sort leaf locations on their supported max height.

**2. Module `stock_storage_type_putaway_abc`:**

For ABC strategy we apply the same sort among on each chunk of locations "A", "B" and "C".

If we have the following locations:

- `LOC_A_MAXHEIGHT_200`
- `LOC_A_MAXHEIGHT_100`
- `LOC_B_MAXHEIGHT_400`
- `LOC_B_MAXHEIGHT_300`
- `LOC_C_MAXHEIGHT_600`
- `LOC_C_MAXHEIGHT_500`

If the ABC storage of a product is flagged `B`, then the proposed locations will be ordered as:
- `LOC_B_MAXHEIGHT_300`
- `LOC_B_MAXHEIGHT_400`
- `LOC_C_MAXHEIGHT_500`
- `LOC_C_MAXHEIGHT_600`
- `LOC_A_MAXHEIGHT_100`
- `LOC_A_MAXHEIGHT_200`

(ordered at first with `BCA` as usual, then by `max_height`)

sebalix committed:
[IMP] stock_storage_type: sort putaway locations by supported max_height

New max_height field on the the location storage type to indicate the
max height supported by a location among all its storage type.

We want to propose the first location corresponding to the height of
a package by losing a minimum of space, so we sort leaf locations on
their supported max height.

sebalix committed:
[FIX] stock_storage_type_putaway_abc: choose B locations instead of A if there is no C

sebalix committed:
[IMP] stock_storage_type_putaway_abc: sort also locations by max heights

For ABC strategy we apply the same sort among on each chunk of
locations "A", "B" and "C".

If we have the following locations:

    LOC_A_MAXHEIGHT_200
    LOC_A_MAXHEIGHT_100
    LOC_B_MAXHEIGHT_400
    LOC_B_MAXHEIGHT_300
    LOC_C_MAXHEIGHT_600
    LOC_C_MAXHEIGHT_500

If the ABC storage of a product is flagged B, then the proposed
locations will be ordered as:

    LOC_B_MAXHEIGHT_300
    LOC_B_MAXHEIGHT_400
    LOC_C_MAXHEIGHT_500
    LOC_C_MAXHEIGHT_600
    LOC_A_MAXHEIGHT_100
    LOC_A_MAXHEIGHT_200

(ordered at first with BCA as usual, then by max_height)

sebalix committed:
[FIX] stock_storage_type_putaway_abc: fix tests to compare list of IDs instead of recordset

guewen commented:
Would be great to have your nice PR's description in the commit's body :)

Commit:
fix error when we have not all kinds (A/B/C) of locations

guewen reviewed (changes requested):
Need https://github.com/sebalix/wms/pull/2 to be merged before

Commit:
Fix sorting broken by a limit on search for valid locations

In this context, "self" are sorted locations. The search returns valid
locations. Later, we filter the locations in "self" by keeping only the
valid locations, but keeping the order of "self.ids".  If we apply a
limit on the intermediate search for valid locations, the order of
"self.ids" may be wrong as the search will return only the first valid
id(s) according to its own order which may be different.

guewen commented:
And https://github.com/sebalix/wms/pull/3

sebalix commented:
@guewen fixing PR merged and messages of commits updated

guewen commented:
/ocabot merge minor
